primitives/npos-elections/src/pjr.rs

```diff
@@ -0,0 +1,508 @@
+		}
+	}
+
+	// Convert Suppports into a SupportMap
+	//
+	// As a flat list, we're limited to linear search. That gives the production of `candidates`,
+	// below, a complexity of `O(s*c)`, where `s == supports.len()` and `c == all_candidates.len()`.
+	// For large lists, that's pretty bad.
+	//
+	// A `SupportMap`, as a `BTreeMap`, has access timing of `O(lg n)`. This means that constructing
+	// the map and then indexing from it gives us timing of `O((s + c) * lg(s))`. If in the future
+	// we get access to a deterministic `HashMap`, we can further improve that to `O(s+c)`.
+	//
+	// However, it does mean allocating sufficient space to store all the data again.
+	let supports: SupportMap<AccountId> = supports.iter().cloned().collect();
+
+	let candidates = all_candidates.into_iter().enumerate().map(|(i, c)| {
+		candidates_index.insert(c.clone(), i);
+
+		// set the backing value and elected flag if the candidate is among the winners.
+		let who = c;
+		let maybe_support = supports.get(&who);
+		let elected = maybe_support.is_some();
+		let backed_stake = maybe_support.map(|support| support.total).unwrap_or_default();
+
+		debug_assert!(
+			elected == (backed_stake > 0),
+			"If a candidate is elected, then it must have a positive backing as well."
+		);
+
+		Candidate { who, elected, backed_stake, ..Default::default() }.to_ptr()
+	}).collect::<Vec<_>>();
+
+	// collect all voters into a unified Vec<Voters>.
+	let voters = all_voters.into_iter().map(|(v, w, ts)| {
+		let mut edges: Vec<Edge<AccountId>> = Vec::with_capacity(ts.len());
+		for t in ts {
+			if edges.iter().any(|e| e.who == t) {
+				// duplicate edge.
+				continue;
+			}
+
+			if let Some(idx) = candidates_index.get(&t) {
+				// if this edge is among the assignments, set the weight as well.
+				let weight = assignment_map
+					.get(&v)
+					.and_then(|d| d.iter().find_map(|(x, y)| if x == &t { Some(y) } else { None }))
+					.cloned()
+					.unwrap_or_default();
+				edges.push(Edge {
+					who: t,
+					candidate: Rc::clone(&candidates[*idx]),
+					weight,
+					..Default::default()
+				});
+			}
+		}
+
+		let who = v;
+		let budget: ExtendedBalance = w.into();
+		Voter { who, budget, edges, ..Default::default() }
+	}).collect::<Vec<_>>();
+
+	(candidates, voters)
+}
+
+/// Check a solution to be t-PJR.
+///
+/// ### Semantics
+///
+/// The t-PJR property is defined in the paper ["Validator Election in Nominated Proof-of-Stake"][NPoS],
+/// section 5, definition 1.
+///
+/// In plain language, the t-PJR condition is: if there is a group of `N` voters
+/// who have `r` common candidates and can afford to support each of them with backing stake `t`
+/// (i.e `sum(stake(v) for v in voters) == r * t`), then this committee needs to be represented by at
+/// least `r` elected candidates.
+///
+/// Section 5 of the NPoS paper shows that this property is equal to: For a feasible solution, if
```

Comment:
While the sentence after the colon is correct, the property isn't equal to anything. The test implies the condition, but not every solution that satifies the condition passes the test.

Reply:
Updated verbiage in 32da70d.